xmrstak/backend/amd/amd_gpu/opencl/wolf-skein.cl

```diff
@@ -22,68 +23,59 @@ static const __constant ulong SKEIN512_256_IV[8] =
 	0xC36FBAF9393AD185UL, 0x3EEDBA1833EDFC13UL
 };
 
-#define SKEIN_INJECT_KEY(p, s)	do { \
+#define SKEIN_INJECT_KEY(p, s, q)	do { \
 	p += h; \
-	p.s5 += t[s % 3]; \
-	p.s6 += t[(s + 1) % 3]; \
-	p.s7 += s; \
+	p.s5 += t[s]; \
+	p.s6 += t[select(s + 1U, 0U, s == 2U)]; \
+	p.s7 += q; \
 } while(0)
 
-ulong SKEIN_ROT(const uint2 x, const uint y)
-{
-	if(y < 32) return(as_ulong(amd_bitalign(x, x.s10, 32 - y)));
-	else return(as_ulong(amd_bitalign(x.s10, x, 32 - (y - 32))));
-}
-
-void SkeinMix8(ulong4 *pv0, ulong4 *pv1, const uint rc0, const uint rc1, const uint rc2, const uint rc3)
+void SkeinMix8(ulong4 *pv0, ulong4 *pv1, const ulong4 rc)
 {
 	*pv0 += *pv1;
-	(*pv1).s0 = SKEIN_ROT(as_uint2((*pv1).s0), rc0);
-	(*pv1).s1 = SKEIN_ROT(as_uint2((*pv1).s1), rc1);
-	(*pv1).s2 = SKEIN_ROT(as_uint2((*pv1).s2), rc2);
-	(*pv1).s3 = SKEIN_ROT(as_uint2((*pv1).s3), rc3);
+	*pv1 = rotate(*pv1, (ulong4)rc);
```

Comment:
I don't understand why a single rotate builtin wasn't used in the first place...

Reply:
I did some digging and it's apparently a quirk of AMD's GCN artitecture which doesn't optimize well with rotate with OpenCL which dates back to Tahiti and earlier. Not sure if that has changed now but you should do some regression testing on different hardware to see if that is still indeed the case.

See also:
https://github.com/magnumripper/JohnTheRipper/issues/1819
http://www.openwall.com/lists/john-dev/2015/10/10/1

Reply:
Which is why I try to test these modifications. I had a small hash rate increase when I patched this. 

Even if rotate doesn’t compile down to a native instruction, the compiler will see a native vector operation rather than that exploded hunk of individual operations that simulate a rotate. The built in at the very least has a macro internally that may beat out what was seen earlier.

Reply:
Additionally, this removes a dependency on one of the amd-only extensions. One step closer to universal opencl code.

Reply:
@simonlui - perhaps use function multi-versioning or a function handle to detect this quirk and apply the appropriate approach at runtime based on the GCN architecture?